tools/protoc-gen-docs/htmlGenerator.go

```diff
@@ -0,0 +1,869 @@
+			}
+		}
+
+		// now, adjust any headers included in the comment to correspond to the right
+		// level, based on the heading level of the surrounding content
+		for i := 0; i < len(lines); i++ {
+			l := lines[i]
+			if strings.HasPrefix(l, "#") {
+				if g.grouping {
+					lines[i] = "##" + l
+				} else {
+					lines[i] = "#" + l
+				}
+			}
+		}
+
+		// find any type links of the form [name][type] and turn
+		// them into normal HTML links
+		for i := 0; i < len(lines); i++ {
+
+			lines[i] = typeLinkPattern.ReplaceAllStringFunc(lines[i], func(match string) string {
+				end := 0
+				for match[end] != ']' {
+					end++
+				}
+
+				linkName := match[1:end]
+				typeName := match[end+2 : len(match)-1]
+
+				if o, ok := g.model.allCoreDescByName["."+typeName]; ok {
+					return g.linkify(o, linkName)
+				}
+
+				if l, ok := wellKnownTypes[typeName]; ok {
+					return "<a href='" + l + "'>" + linkName + "</a>"
+				}
+
+				return "*" + linkName + "*"
+			})
+		}
+	}
+
+	text = strings.Join(lines, "\n")
+
+	// turn the comment from markdown into HTML
+	result := blackfriday.Run([]byte(text))
+
+	g.buffer.Write(result)
+	g.buffer.WriteByte('\n')
+}
+
+// well-known types whose documentation we can refer to
+var wellKnownTypes = map[string]string{
+	"google.protobuf.Duration":               "https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#google.protobuf.Duration",
+	"google.protobuf.Timestamp":              "https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#google.protobuf.Timestamp",
+	"google.api.MetricDescriptor.MetricKind": "https://cloud.google.com/logging/docs/reference/v2/rest/v2/projects.metrics#metrickind",
+	"google.api.MetricDescriptor.ValueType":  "https://cloud.google.com/logging/docs/reference/v2/rest/v2/projects.metrics#valuetype",
+	"google.api.MetricDescriptor":            "https://cloud.google.com/logging/docs/reference/v2/rest/v2/projects.metrics#metricdescriptor",
+}
+
+func (g *htmlGenerator) linkify(o coreDesc, name string) string {
+	if o == nil {
+		return name
+	}
+
+	if msg, ok := o.(*messageDescriptor); ok && msg.GetOptions().GetMapEntry() {
+		return name
+	}
+
+	known := wellKnownTypes[g.absoluteTypeName(o)]
+	if known != "" {
+		return "<a href='" + known + "'>" + name + "</a>"
+	}
+
+	loc := o.fileDesc().parent.location
+	fragment := "#" + dottedName(o)
+
+	if loc != "" && loc != g.currentFile.parent.location {
+		return "<a href='" + loc + fragment + "'>" + name + "</a>"
+	}
+
+	return "<a href='" + fragment + "'>" + name + "</a>"
+}
+
+func (g *htmlGenerator) warn(loc *descriptor.SourceCodeInfo_Location, format string, args ...interface{}) {
+	if g.genWarnings {
+		place := ""
+		if loc != nil {
+			place = fmt.Sprintf("%s:%d:%d:", g.currentFile.GetName(), loc.Span[0], loc.Span[1])
+		}
+
+		fmt.Fprintf(os.Stderr, place+" "+format+"\n", args...)
+	}
+}
+
+func (g *htmlGenerator) relativeTypeName(obj coreDesc) string {
+	typeName := dottedName(obj)
```

Comment:
is there a way to special case enums and make them be small letters instead of the huge CAPS that gets displayed?

Reply:
Let's land this end-to-end and get istio.io up to date, and then we can easily do the kind of tuning you are proposing. I think we have to make these docs excellent, and there's a lot of room for improvement considering what we have today.